src/main/java/com/android/volley/AsyncNetwork.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.volley;
+
+import androidx.annotation.Nullable;
+import androidx.annotation.RestrictTo;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class AsyncNetwork implements Network {
+
+    public interface OnRequestComplete {
+        void onSuccess(NetworkResponse networkResponse);
+
+        void onError(VolleyError volleyError);
+    }
+
+    public abstract void performRequest(Request<?> request, OnRequestComplete callback);
+
+    @Override
+    public NetworkResponse performRequest(Request<?> request) throws VolleyError {
+        // Is there a better way to go about this?
```

Comment:
looks mostly fine to me, but personally I'd probably scrap the Response POJO and just have two AtomicReferences, one for NetworkResponse and one for VolleyError. It doesn't seem like the container is buying you much.

Reply:
Okay, that's what I had, but switched it to keep the same structure as AsyncHttpStack.